cdsodatacli/query.py

```diff
@@ -279,126 +288,84 @@ def urls_creat(gdf, top=None):
         url = (urlapi + str_query + '&$expand=Attributes')
-        urls.append(url)
-
+        urls.append((enter_index, url))
     end_time = time.time()
     processing_time = end_time - start_time
-    print(f"creat_urls processing time:{processing_time}s")
+    logging.debug(f"creat_urls processing time:{processing_time}s")
     return urls
 
 
-def fetch_data_from_urls(urls):
-    start_time = time.time()
-    collected_data = []
-
-    with tqdm(total=len(urls)) as pbar:
-        for url in urls:
-            json_data = requests.get(url).json()
-            if 'value' in json_data:
-                collected_data = pd.DataFrame.from_dict(json_data['value'])
-                pbar.update(1)
-
-    collected_data = pd.DataFrame(collected_data)
-    end_time = time.time()
-    processing_time = end_time - start_time
-    print(f"fetch_data_from_urls time:{processing_time}s")
-    return collected_data
-
-
-#def fetch_url(url):
-#    data = requests.get(url).json()
-#    return process_data(data)
-
-
-#def fetch_data_from_urls(urls):
-#    with ThreadPoolExecutor(max_workers=10) as executor:
-#        data = list(tqdm(executor.map(fetch_url, urls), total=len(urls)))
-
-#    df = pd.concat(data, ignore_index=True)
-#    return df
+def get_cache_filename(url):
+    url_hash = hashlib.md5(url.encode('utf-8')).hexdigest()
+    return os.path.join(cache_dir, url_hash + '.json')
 
 
-def process_data(json_data):
-    """
-    Processes the fetched JSON data and returns relevant information.
-
-    :param json_data: JSON data containing the fetched results.
-    :return: Processed data for visualization.
-    """
-    res = None
-    if 'value' in json_data:
-        res = pd.DataFrame.from_dict(json_data['value'])
+def fetch_data_from_url(url, index):
+    cache_file = get_cache_filename(url)
+    if os.path.exists(cache_file):
+        with open(cache_file, 'r') as f:
+            data = json.load(f)
     else:
-        print("No data found.")
-        pass
-    return res
+        response = requests.get(url)
+        data = response.json()
+        if 'value' in data:
+            with open(cache_file, 'w') as f:
+                json.dump(data, f)
+    # process data
+    df = pd.DataFrame.from_dict(data['value'])
+    df['input_index'] = index
+    return df
+
+
+def fetch_data_from_urls(urls, max_workers=50):
+    collected_data = pd.DataFrame()
+
+    with ThreadPoolExecutor(max_workers=max_workers) as executor, tqdm(total=len(urls)) as pbar:
+        future_to_url = {executor.submit(fetch_data_from_url, url, index): (url, index) for index, url in urls}
+        for future in as_completed(future_to_url):
+            df = future.result()
+            collected_data = pd.concat([collected_data, df])
+            pbar.update(1)
+    return collected_data
 
 
-def remove_duplicates(safes_ori, keep_list=[]):
-    """
-    Remove duplicate safe (ie same footprint with same date, but different prodid)
-    """
+def remove_duplicates(safes):
```

Comment:
could you add docstrings to your methods?

Reply:
Alright